### DISCUSSION TITLE
"x86 Netbook BareMetal" TIC-80 Project. Where to start?

### DISCUSSION THREAD

**Discussion by peeweek** [nesbox/TIC-80#2182] [Q&A]
Hello there, 

I own (since quite a long time) an [Asus EEEPC 1015](https://www.notebookcheck.net/Asus-Eee-PC-1015-Series.35259.0.html) 10 inch netbook that is not about any use for me. Lately I've been wondering about repurposing it for my son to learn programming on some kind of "Tiny Computer", and especially on TIC-80.

I have tried to install a base debian, with some kind of autologin, and autostart X, then TIC-80. But it seems that the startup takes forever and feels clunky. 

Then I stumbled upon many things around: first the PICO-8 RPI port that was using directfb instead of X to run in a bare metal fashion, and also the same happening to TIC-80... but still for Raspberry Pi.

I wonder if such a project could also be feasible on x86 so I can boot almost instantly on TIC80 and it would shutdown the computer once it exists. 

I think that probably using some kind of lightweight distro (puppy, arch, or even linux lite), but then I have absolutely no clue on how to compile a build of pico8 that would run out of X11, maybe using directFB. And more importantly how to be able to bundle everything once my prototype runs so I can share it with others.

So far, the only documentation I found is the [directfb readme](https://github.com/DirectFB/directfb) but I have no clue on how to modify TIC-80 so it can run with directfb instead of X.

Does someone have experience with it to give me directions on how to start?

Thanks,

**Comment by Anrock** [comment]
Booting up x86 and rpi from nothing are pretty different things so TIC-80 on bare x86 would be a project on it's own, I believe.

I would keep looking at minimal distros in your place. While EEEPC 1015 is a potato I've used it before for light web browsing and developing pet projects so it definitely should be able to run TIC-80 with good performance and low latency.

**Comment by bztsrc** [comment]
> I think that probably using some kind of lightweight distro (puppy, arch, or even linux lite), but then I have absolutely no clue on how to compile a build of pico8 that would run out of X11, maybe using directFB.

You don't need a distro for that. All you need is a Linux kernel that runs TIC-80 as a single `init` process. But you'll have to patch TIC-80 to use directFB because there'll be no X11.

See this project for details: https://github.com/shadlyd15/DoomLinux This runs fbdoom as a single process, so just replace that with tic80 and modify init to start tic80 instead of fbdoom (here `init` is a shell script that does all the necessary mounting and then executes a command).

> I have tried to install a base debian, with some kind of autologin, and autostart X, then TIC-80. But it seems that the startup takes forever and feels clunky.

You should be able to get this distro-less approach with X11 too; just add the libX11 libraries and basic X11 commands to rootfs.gz, call `startx` from init, and put tic80 invocation in `xinitrc`. This needs a little more work, but in exchange you don't have to patch TIC-80 for fbdev. I'm pretty sure it will boot much much faster than any distro using systemd, autologin, etc. (Also, by compiling a static kernel for EEEPC you'll eliminate unnecessary modules, dynamic module loading altogether, so your kernel will be smaller and will include everything it needs right away, therefore will boot much faster.)

> Booting up x86 and rpi from nothing are pretty different things so TIC-80 on bare x86 would be a project on it's own, I believe.

Definitely. The bare metal RPi port does not use any kernel, it uses a hack with the low-level, hardware-specific Circle library instead. Very different to x86.

Hope this helps,
bzt
